processing/src/main/java/org/apache/druid/frame/processor/FrameProcessorExecutor.java

```diff
@@ -0,0 +1,602 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.frame.processor;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Multimaps;
+import com.google.common.collect.SetMultimap;
+import com.google.common.collect.Sets;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
+import it.unimi.dsi.fastutil.ints.IntSet;
+import org.apache.druid.frame.channel.ReadableFrameChannel;
+import org.apache.druid.frame.channel.WritableFrameChannel;
+import org.apache.druid.java.util.common.Either;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.java.util.common.concurrent.Execs;
+import org.apache.druid.java.util.common.guava.Sequence;
+import org.apache.druid.java.util.common.io.Closer;
+import org.apache.druid.java.util.common.logger.Logger;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+
+/**
+ * Manages execution of {@link FrameProcessor} in an {@link ExecutorService}.
```

Comment:
Explain a bit more?

This seems to assume that each "fragment" of execution is handled within a single frame processor: reads from (input | network), writes to (destination | output).

Yet, most pipelines are made up of multiple operations: scan, filter and project, say. Or, aggregate & filter. Etc.

If each "fragment" must do multiple operations, does that mean that every combination of pipeline elements has to be code up as a different frame processor? And that the developer has to know to route the one-frame-per-call logic to the step in that pipeline? (That is, route to the scan operation, then do a filter, and stop. Next call, pick up that frame and write to output and stop, etc.)

Or, is the intent here that a "fragment" (unit of pipeline execution running on a node) is a combination of frame processors, and this class handles multiplexing calls to the elements of the pipeline? Call the scan. When it has a frame in its output buffer, call the filter. If it has an output frame, call write. Circle back ground to the scan otherwise.

Would be great if we could explain this pipeline-level architecture somewhere.

Reply:
Hmm. The way I'd imagined it is that each of these "fragments" would be composed of:

- a processor for gathering inputs (for example: if reading sorted data, this would be a merge-sorting processor).
- a processor with the main logic, which could internally contain multiple logical operators: it could do both a filter and a project, for example.
- processor(s) for preparing outputs (for example: if writing sorted data, this would be a SuperSorter; if doing hash partitioning, it would be a hash-partitioning processor).

I didn't imagine a separate processor for each logical operator, because I thought it would be too inefficient to have to create a frame in between each logical operator.

Rather, I'd imagined that we would introduce operators that get "packed into" processors, and that transfer data between themselves in some more efficient way than creating frames.

Reply:
That makes sense. So, to run the "fragment" you just described, we'd have something like this?

```text
reader processor --> channel --> main logic processor --> channel --> output processor
```

In this case, there are three concurrent frame processors communicating over in-memory channels? If so, it is very Go-like! This leads to two questions.

First, how does the "runner" know what to run? A comment in the sorter suggests we'd prefer to do the output, then do the main logic that can produce more output, and finally do a read only if we've used up whatever data we have. Clear enough.

How does the runner coordinate? What is the protocol that each frame processor uses to say "I can run" or "I could run, if you gave me more input" or "Sorry, EOF, I'm fresh out of data?

And, in the main logic, what structure, if any, do we provide if that logic is, say, a filter, projection and frame-repackaging? How do the various steps coordinate?

This is the kind of nitty-gritty detail that each frame processor writer will need to know. Not sure where that info should go, but it would be handy if we could explain it.

Reply:
> So, to run the "fragment" you just described, we'd have something like this?
>
> reader processor --> channel --> main logic processor --> channel --> output processor
>
> In this case, there are three concurrent frame processors communicating over in-memory channels? If so, it is very Go-like! 

Yeah, that's exactly how I'd imagined it.

> How does the runner coordinate? What is the protocol that each frame processor uses to say "I can run" or "I could run, if you gave me more input" or "Sorry, EOF, I'm fresh out of data?

This is what the ReturnOrAwait is for:

- "I can run" is `ReturnOrAwait.runAgain()`
- "I could run, if you gave me more input" is `ReturnOrAwait.awaitAny({set})`
- "EOF" is `ReturnOrAwait.returnObject({return value})`

> And, in the main logic, what structure, if any, do we provide if that logic is, say, a filter, projection and frame-repackaging? How do the various steps coordinate?

I think that's TBD. In the near future (next patch in this sequence) I'm imagining that we actually don't have operators yet. Instead, each processor corresponds to a piece of a native query, and they're each written bespoke. One of the goals of the operator project would be to introduce the structure you are asking about.